accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -1461,3 +1461,319 @@ wrapper class is so huge that it is likely to be a major use case for
+whether it’s the `true` or the `false` value, which is still sound. _I do not
+believe we make any assumptions about members of `bool`, or invoke any
+explicitly._
+
+#### Numbers and arithmetic operators
+
+We have special rules in type inference for number operations on `int`, `num`
+and `double`, to ensure that the static type of `1 + 1` is `int`. We effectively
+pretend to have overloaded operators for typing only.
+
+The [newest
+version](https://github.com/dart-lang/language/blob/main/accepted/2.12/nnbd/number-operation-typing.md)
+of those rules came with null safety. The way the rules are written, they apply,
+e.g., to `e1 + e2` when the static type of `e1` is a non-`Never` subtype of
+`int` (the receiver is an `int`, maybe typed as a type variable `X` with bound
+`int`), and the static type of `e2` is a subtype of `int`. That needs to also
+say that the `+` must denote an instance member, to avoid applying to an
+extension type implementing `int` and declaring an extension type  `+` operator.
+
+That amounts to two sections being changed by inserting something to the effect
+of “where it’s an instance member”, additions italicized here:
+
+>Let `e` be an expression of one of the forms `e1 + e2`, `e1 - e2`, `e1 * e2`,
+>`e1 % e2` or `e1.remainder(e2)`, where the static type of `e1` is a
+>non-~~`Never`~~_bottom_ type *T* and *T* <: `num`, ~~and~~ where the static
+>type of `e2` is *S* and *S* is assignable to `num`_, and where *T* has an
+>implementation of the operator, or `remainder` method, that is an instance
+>member_. Then:
+
+and
+
+>Let `e` be a normal invocation of the form `e1.clamp(e2, e3)`, where the static
+>types of `e1`, `e2` and `e3` are *T*<sub>1</sub>, *T*<sub>2</sub> and
+>*T*<sub>3</sub> respectively, ~~and~~ where *T*<sub>1</sub>, *T*<sub>2</sub>,
+>and *T*<sub>3</sub> are all non-~~`Never`~~_bottom_ subtypes of `num`_, and
+>where *T*<sub>1</sub> has an instance member named `clamp`_ . Then:
+
+This will prevent the rules from applying to invocations of extension type
+operators or methods, which shadow the platform methods that we know and trust.
+
+No other changes should be needed. The remainder of the rules are about context
+types, and require the types involved to be *supertypes* of a number interface
+type, which extension types never are.
+
+#### Patterns
+
+All map pattern entry patterns and list pattern element patterns are define in
+terms of calling `length,`, `operator[]` and/or `containsKey` on the object. All
+these invocations *must* be instance member invocations, whether the static type
+of the matched value type is an extension type implementing `Map` or `List` or
+not, and whether it defines its own `length`, `operator[]` or `containsKey`
+extension type members. For all practical purposes, the value is cast to `Map<K,
+V>`/`List<E>` for some `K` and `V`, or `E`, *before* accessing elements, and it
+uses the caching behavior of instance elements.
```

Comment:
I'd prefer if we avoid the "magic" and simply make it a compile-time error for a collection pattern to invoke `length` etc. if it is an extension type instance member rather than a class instance member.

Next, I'm not 100% convinced that it should be an error. It could be a useful hook.

Just talked about this IRL: We might consider allowing extension type instance members and/or extension instance members to provide the required members to support iteration:

So the requirement for being usable in a `for` loop could be (1) must be an `Iterable<T>` for some `T` (the static type could be `dynamic` or some `S <: Iterable<T>`), or (2) must have a static type which is an extension type that has the required members (`iterator`, yielding an `Iterator<T>` for some `T`, or returning an object typed as `E` which is an extension type that has `moveNext()` and `current`).

We might want to treat `for` loops differently than other built-in constructs, because they are so performance critical. (For instance, it might be very important that we could perhaps use a native kind of iteration on a `JSArray`.)

Reply:
> I'd prefer if we avoid the "magic" and simply make it a compile-time error for a collection pattern to invoke length etc. if it is an extension type instance member rather than a class instance member.

That could apply to collection patterns in declaration patterns, but not in refutable patterns.

That is, it's a compile-time error if the matched value type of an irrefutable list pattern (in a declaration/assignment pattern) has an extension type instance member named `length` or `operator[]`, and if the matched value type of an irrefutable map pattern has an extension type instance member named `contains` or `operator[]`.

Or we can say that the list/map patterns always perform class instance member invocations for those methods, which is possibly what it says today. (But most likely it says nothing conclusive, because there used to be nothing to say.)